proxy/healthy_endpoints_test.go

```diff
@@ -128,63 +151,135 @@ func TestPHCForSingleHealthyEndpoint(t *testing.T) {
+			va := fireVegeta(t, ps, 3000, 1*time.Second, 5*time.Second)
+			count200, ok := va.CountStatus(200)
+			assert.True(t, ok)
+			assert.Equal(t, count200, 15000) // 3000*5s
+			assert.Equal(t, count200, int(va.TotalRequests()))
 		})
 	}
 
 	t.Run("consistent hash with balance factor", func(t *testing.T) {
 		_, ps := setupProxy(t, fmt.Sprintf(`* -> consistentHashKey("${request.header.ConsistentHashKey}") -> consistentHashBalanceFactor(1.25) -> <consistentHash, "%s", "%s", "%s">`,
-			services[0].URL, services[1].URL, services[2].URL))
-		failedReqs := sendGetRequests(t, ps)
-		assert.Equal(t, 0, failedReqs)
+			services[0], services[1], services[2]))
+		va := fireVegeta(t, ps, 3000, 1*time.Second, 5*time.Second)
+		count200, ok := va.CountStatus(200)
+		assert.True(t, ok)
+		assert.Equal(t, count200, int(va.TotalRequests()))
 	})
 }
 
 func TestPHCForMultipleHealthyAndOneUnhealthyEndpoints(t *testing.T) {
-	services := []*httptest.Server{}
-	for i := 0; i < 3; i++ {
-		serviceNum := i
-		service := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			if serviceNum == 0 {
-				// emulating unhealthy endpoint
-				time.Sleep(100 * time.Millisecond)
-			}
-			w.WriteHeader(http.StatusOK)
-		}))
-		services = append(services, service)
-		defer service.Close()
-	}
-
-	for _, algorithm := range []string{"random", "consistentHash", "roundRobin", "powerOfRandomNChoices"} {
+	services := setupServices(t, 2, 1)
+	for _, algorithm := range []string{"random", "roundRobin", "powerOfRandomNChoices"} {
 		t.Run(algorithm, func(t *testing.T) {
 			mockMetrics, ps := setupProxy(t, fmt.Sprintf(`* -> backendTimeout("5ms") -> consistentHashKey("${request.header.ConsistentHashKey}") -> <%s, "%s", "%s", "%s">`,
-				algorithm, services[0].URL, services[1].URL, services[2].URL))
-			failedReqs := sendGetRequests(t, ps)
-			assert.InDelta(t, 0, failedReqs, 0.1*float64(nRequests))
+				algorithm, services[0], services[1], services[2]))
+
+			va := fireVegeta(t, ps, 3000, 1*time.Second, 5*time.Second)
+
+			count200, ok := va.CountStatus(200)
+			assert.True(t, ok)
+
+			count504, ok := va.CountStatus(504)
+			assert.True(t, ok)
+
+			failedRequests := math.Abs(float64(va.TotalRequests())) - float64(count200)
+			t.Logf("total requests: %d, count200: %d, count504: %d, failedRequests: %f", va.TotalRequests(), count200, count504, failedRequests)
+
+			assert.Equal(t, float64(count504), failedRequests)
+			assert.InDelta(t, 0, float64(failedRequests), 0.1*float64(va.TotalRequests()))
 			mockMetrics.WithCounters(func(counters map[string]int64) {
-				assert.InDelta(t, float64(nRequests), float64(counters["passive-health-check.endpoints.dropped"]), 0.3*float64(nRequests)) // allow 30% error
+				assert.InDelta(t, float64(va.TotalRequests()), float64(counters["passive-health-check.endpoints.dropped"]), 0.6*float64(va.TotalRequests())) // allow 60% error
 			})
 		})
 	}
 
+	t.Run("consistentHash", func(t *testing.T) {
```

Comment:
ConsistentHash was separated because it's a little bit tricky to set headers for different set of requests with Vegeta, we use the hash to map the request to endpoint.